File: manual/code/test_suites/cpp/FunctionBuildTest/src/function_numpy_divide.cpp
#include <stdexcept>
#include <typeinfo>

class DivisionCalculator {
public:
    static double divide(const Number& arr1, const Number& arr2, const Number& scalar) {
        class InnerCalculator {
        public:
            double divideArrays(const Number& a1, const Number& a2) {
                if (a2.isNull()) {
                    throw std::invalid_argument("arr2 must not be null for this operation.");
                }
                if (!a1.isNumber() || !a2.isNumber()) {
                    throw std::invalid_argument("Both arr1 and arr2 must be basic data types (int or float).");
                }
                return a1.toDouble() / a2.toDouble();
            }

            double divideByScalar(const Number& a1, const Number& scalarValue) {
                if (scalarValue.isNull()) {
                    throw std::invalid_argument("Scalar must not be null for this operation.");
                }
                if (!a1.isNumber()) {
                    throw std::invalid_argument("arr1 must be a basic data type (int or float) when dividing by scalar.");
                }
                return a1.toDouble() / scalarValue.toDouble();
            }
        };

        InnerCalculator calculator;

        if (!arr2.isNull()) {
            return calculator.divideArrays(arr1, arr2);
        } else if (!scalar.isNull()) {
            return calculator.divideByScalar(arr1, scalar);
        } else {
            throw std::invalid_argument("Either arr2 or scalar must be provided.");
        }
    }
};

File: manual/code/generate_strategies/IR(summary)/python/python/function_numpy_divide.py
import numpy as np


class DivisionCalculator:

    def divide(self, arr1, arr2=None, scalar=None):
        try:
            if arr2 is not None:  
                
                if not isinstance(arr1, (int, float)) or not isinstance(arr2, (int, float)):
                    raise ValueError("Both arr1 and arr2 must be basic data types (int or float).")

                
                result = arr1 / arr2
                return result  

            elif scalar is not None:  
                if not isinstance(arr1, (int, float)):
                    raise ValueError("arr1 must be a basic data type (int or float) when dividing by scalar.")

                
                result = arr1 / scalar
                return result  

            else:
                raise ValueError("Either arr2 or scalar must be provided.")

        except Exception as e:
            raise ValueError(f"Error in division: {e}")